net: lib: dhcpv4: add possibility to parse multiple DNS servers received from DHCP server

AndreyDodonov-EH opened the pull request:
Currently if we receive multiple DNS servers via DHCP options, only the first one is used,
regardless of CONFIG_DNS_RESOLVER_MAX_SERVERS constant.

Parse multiple servers and save as many DNS server addresses as fits in CONFIG_DNS_RESOLVER_MAX_SERVERS

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
 
-			if (net_pkt_read(pkt, dns.sin_addr.s4_addr, 4) ||
-			    net_pkt_skip(pkt, length - 4U)) {
+			const uint8_t provided_servers_cnt = length/addr_size;
```
Please add space before and after "/"

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
-			const struct sockaddr *dns_servers[] = {
-				(struct sockaddr *)&dns, NULL
-			};
+			struct sockaddr_in dnses[CONFIG_DNS_RESOLVER_MAX_SERVERS];
```
We could do
```
struct sockaddr_in dnses[CONFIG_DNS_RESOLVER_MAX_SERVERS] = { 0 };
```
and avoid the memset() call below.

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+					i < CONFIG_DNS_RESOLVER_MAX_SERVERS; i++) {
+					dns_servers[i] = NULL;
+				}
+				dns_servers_cnt = provided_servers_cnt;
```
Please put empty line after ending `}`

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			for (uint8_t i = 0; i < dns_servers_cnt; i++) {
+				if (net_pkt_read(pkt, dnses[i].sin_addr.s4_addr, addr_size)) {
+					NET_ERR("options_dns, short packet");
+				return false;
```
Indentation

jukkar commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+				}
+			}
+
+			if (net_pkt_skip(pkt, length - dns_servers_cnt*addr_size)) {
```
Please add spaces around `*`

jukkar reviewed (commented):
LGTM, just very minor style nits

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+				dns_servers_cnt = CONFIG_DNS_RESOLVER_MAX_SERVERS;
+			} else {
+				for (uint8_t i = provided_servers_cnt;
+					i < CONFIG_DNS_RESOLVER_MAX_SERVERS; i++) {
```
Please align with the opening bracket, so that the if condition does not blend with the code block.

rlubos commented on subsys/net/lib/dhcpv4/dhcpv4.c:
```diff
@@ -1056,33 +1056,59 @@ static bool dhcpv4_parse_options(struct net_pkt *pkt,
+			for (uint8_t i = 0; i < CONFIG_DNS_RESOLVER_MAX_SERVERS; i++) {
+				dns_servers[i] = (struct sockaddr *)&dnses[i];
+			}
 			int status;
```
nit, we typically keep variable definitions on top of the block, same below.

AndreyDodonov-EH committed:
net: lib: dhcpv4: parse multiple DNS servers received from DHCP

If we receive multiple DNS servers via DHCP only the first one is used,
regardless of CONFIG_DNS_RESOLVER_MAX_SERVERS constant.
Parse DHCP option and save addresses that fit

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>
